tests/sidewinder/working/s2n-cbc/patches/cbc.patch

```diff
@@ -0,0 +1,86 @@
+diff --git a/tests/sidewinder/working/s2n-cbc/tls/s2n_cbc.c b/tests/sidewinder/working/s2n-cbc/tls/s2n_cbc.c
+index b37efb0..077d214 100644
+--- a/tests/sidewinder/working/s2n-cbc/tls/s2n_cbc.c
++++ b/tests/sidewinder/working/s2n-cbc/tls/s2n_cbc.c
+@@ -25,6 +25,28 @@
+ 
+ #include "tls/s2n_record.h"
+ 
++#include <smack.h>
++#include <smack-contracts.h>
++#include "ct-verif.h"
++#include "sidewinder.h"
++
++/* Because of a bug in SMACK, when we try to specify the invarients for this loop in the main function, 
++ * SMACK crashes. Moving the code (unmodified) to this function sidesteps the bug. Michael Emmi has promissed
++ * a fix soon
++ */
++int double_loop(int old_mismatches, struct s2n_blob *decrypted, int check, int cutoff, int padding_length) {
++  __VERIFIER_assert(decrypted->size >= 0);
++  __VERIFIER_assert(decrypted->size <= 1024);
++  int mismatches = old_mismatches;
++  for (int i = 0, j = decrypted->size - 1 - check; i < check && j < decrypted->size; i++, j++) {
++    invariant(i <= check);
++    invariant(j == i + decrypted->size - check - 1);
++    uint8_t mask = ~(0xff << ((i >= cutoff) * 8));
++    mismatches |= (decrypted->data[j] ^ padding_length) & mask;
++  }
++  return mismatches;
++}
++
+ /* A TLS CBC record looks like ..
+  *
+  * [ Payload data ] [ HMAC ] [ Padding ] [ Padding length byte ]
+@@ -53,17 +75,20 @@ int s2n_verify_cbc(struct s2n_connection *conn, struct s2n_hmac_state *hmac, str
+        copy = &conn->server->record_mac_copy_workspace;
+     }
+     
+-    uint8_t mac_digest_size;
+-    GUARD(s2n_hmac_digest_size(hmac->alg, &mac_digest_size));
++    uint8_t mac_digest_size = DIGEST_SIZE;;
++    //GUARD(s2n_hmac_digest_size(hmac->alg, &mac_digest_size));
+ 
+     /* The record has to be at least big enough to contain the MAC,
+      * plus the padding length byte */
+     gt_check(decrypted->size, mac_digest_size);
+ 
+     int payload_and_padding_size = decrypted->size - mac_digest_size;
++    __VERIFIER_assert(payload_and_padding_size <= MAX_SIZE - 20);
+ 
+     /* Determine what the padding length is */
+     uint8_t padding_length = decrypted->data[decrypted->size - 1];
++    __VERIFIER_assume(padding_length >= 0);
```

Comment:
I like these `__VERIFIER_assume()` calls! 

I think we can (should?) use them in the normal s2n codebase. You are already converting `lte_check()` and `gte_check()` calls to `__VERIFIER_assume()` calls in `tests/sidewinder/working/s2n-cbc/patches/safety.patch`, would it make sense to convert these into `lte_checks()` in the normal codebase and do the same conversion you're already doing so that we get the benefit in both places?

If you just want to open a backlog issue for it though, that would be okay with me too.

Reply:
In this case, the assumes are necessary because Sidewinder is using integer arithmetic, and needs to be informed about the bounds on a uint8 (otherwise, it would pessimistically assume it could have any integer value). So the assertion would always pass in the original code.